### PR TITLE
[GridBundle] Allow filtering by fields from associated objects

### PR DESCRIPTION
**Pull request by aramalipoor**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Related tickets | mentioned in #5476
| License         | MIT

I also had some issues when filter form was submitting with empty string value so I thought maybe do not restrict DataSource when filer data is empty?!

### CODE REVIEW

File: src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php [hunk 1]
```diff
@@ -94,31 +94,39 @@ public function notEquals($field, $value)
...
 
     /**
      * {@inheritdoc}
      */
     public function lessThanOrEqual($field, $value)
     {
-        $this->queryBuilder->andWhere($this->getFieldName($field).' =< :'.$field)->setParameter($field, $value);
+        $this->setParameter($field, $value);
+
+        $this->queryBuilder->andWhere($this->getFieldName($field).' =< :'.$field);
```
Comment (bigfoot90) [inline src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php]: This should be ```<=```

File: src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php [hunk 2]
```diff
@@ -94,31 +94,39 @@ public function notEquals($field, $value)
...
 
     /**
      * {@inheritdoc}
      */
     public function greaterThanOrEqual($field, $value)
     {
-        $this->queryBuilder->andWhere($this->getFieldName($field).' => :%s'.$field)->setParameter($field, $value);
+        $this->setParameter($field, $value);
+
+        $this->queryBuilder->andWhere($this->getFieldName($field).' => :%s'.$field);
```
Comment (bigfoot90) [inline src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php]: This should be ```>=```

**Comment by liverbool** [comment]
@pjedrzejewski why this PR not merge? What do you think?

File: src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php [hunk 3]
```diff
@@ -74,7 +74,7 @@ public function comparison($field, $operator, $value)
      */
     public function equals($field, $value)
     {
-        $this->queryBuilder->setParameter($field, $value);
+        $this->setParameter($field, $value);
 
         return $this->queryBuilder->expr()->eq($this->getFieldName($field), ':'.$field);
```
Comment (liverbool) [inline src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php]: BUG: `$field` still not replace with `_`

File: src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php [hunk 4]
```diff
@@ -84,7 +84,7 @@ public function equals($field, $value)
      */
     public function notEquals($field, $value)
     {
-        $this->queryBuilder->setParameter($field, $value);
+        $this->setParameter($field, $value);
 
         return $this->queryBuilder->expr()->neq($this->getFieldName($field), ':'.$field);
```
Comment (liverbool) [inline src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php]: same

File: src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php [hunk 5]
```diff
@@ -94,31 +94,39 @@ public function notEquals($field, $value)
      */
     public function lessThan($field, $value)
     {
-        $this->queryBuilder->andWhere($this->getFieldName($field).' < :'.$field)->setParameter($field, $value);
+        $this->setParameter($field, $value);
+
+        $this->queryBuilder->andWhere($this->getFieldName($field).' < :'.$field);
```
Comment (liverbool) [inline src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php]: same :0

File: src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php (hunk 3 above)
Comment (aramalipoor) [inline src/Sylius/Bundle/GridBundle/Doctrine/ORM/ExpressionBuilder.php]: OOPS!! MY BAD!!

**Comment by pjedrzejewski** [comment]
Build is failing, otherwise looks good! 👍

**Comment by aramalipoor** [comment]
@pjedrzejewski The fail is about `it filters data containing empty strings` which seems legit, and now I don't know what to decide here :D ?

The  problem is I do not want to apply any filter when form input of filter is not filled by user (e.g. filling only another field) because it will automatically only show data with empty strings! But on the other hand what to do if user **really** wants to filter empty strings?

**Comment by pjedrzejewski** [comment]
The problems is that string filter type: empty / non-empty has no value - so your ``if (empty...`` breaks the tests. I guess we should check if filter type is not these 2.

**Comment by aramalipoor** [comment]
@pjedrzejewski Thanks for the feedback :) I think it's ready now

**Comment by pjedrzejewski** [comment]
Thanks Aram! Tests FTW!

**Pull request merged**